File: capitolo_1/cpp/Esempio1_17.cpp
#include <iostream>
#include <vector>
#include <cmath>
#include <string>
#include <map>
#include "matplotlibcpp.h"
#include "myDiffEquation.h"

namespace plt = matplotlibcpp;

// --------------------------
// Parametri di sistema
// --------------------------
struct Params {
    double Mbar;
    double hbar;
    double kbar;
    double Fe;
    double sbar;
    double mu;
    double nu;
};

// --------------------------
// Equazione differenziale - Controllore ad anello chiuso
// --------------------------
std::vector<double> f_anelloChiuso(double t, const std::vector<double>& x, const Params& p, double /*unused*/) {
    (void)t; // unused
    std::vector<double> dx(3);
    double error = p.sbar - x[0];
    dx[0] = x[1];
    dx[1] = -(p.hbar/p.Mbar) * x[1] -(p.kbar/p.Mbar) * x[0] + (p.mu/p.Mbar) * error + (p.nu/p.Mbar) * x[2] - p.Fe/p.Mbar;
    dx[2] = error;
    return dx;
}

// --------------------------
// MAIN
// --------------------------
int main() {
    // Dati
    double Mbar = 1, hbar = 3, kbar = 1, sbar = 1, Fe = 0;
    std::vector<double> mus_prop = {1.0, 1.5, 3.0};
    std::vector<double> nus_prop = {0.5, 1.0};
    double dt = 0.001;
    double t0 = 0, tf = 10.0;
    int N = static_cast<int>((tf - t0) / dt);
    std::vector<double> time(N + 1);
    for (int i = 0; i <= N; ++i) time[i] = t0 + i * dt;

    // -----------------------------------------
    // Figura 1.12 - Transitorio posizione massa-molla, C.I. nulle, sotto azione del controllore.
    // -----------------------------------------
    std::map<std::pair<double, double>, std::vector<double>> transitorioPosizione;
    std::map<std::pair<double, double>, std::vector<double>> transitorioFmotrice;

    for (auto mu : mus_prop) {
        for (auto nu : nus_prop) {
            Params p{Mbar, hbar, kbar, Fe, sbar, mu, nu};
            double dummy = 0.0; // parametro non usato nella funzione
            auto traj = rungeKutta4(f_anelloChiuso, {0.0, 0.0, 0.0}, t0, tf, dt, p, dummy);
            // Estrae la posizione
            std::vector<double> posizione;
            for (auto& x : traj) posizione.push_back(x[0]);
            transitorioPosizione[std::make_pair(mu, nu)] = posizione;
            // Estrae la forza motrice
            std::vector<double> Fmotrice;
            for (auto& x : traj) Fmotrice.push_back(mu * sbar - mu * x[0] + nu * x[2]);
            transitorioFmotrice[std::make_pair(mu, nu)] = Fmotrice;
        }
    }

    // Plot Figura 1.15a - Transitorio posizione
    plt::figure(1);
    for (auto& [params, C] : transitorioPosizione)
        plt::plot(time, C, {{"label", "μ = " + std::to_string(params.first) + ", ν = " + std::to_string(params.second)}});
    plt::xlabel("Tempo (t)");
    plt::ylabel("Posizione s");
    plt::title("Transitorio della posizione - C.I. nulle");
    plt::legend();
    plt::grid(true);

    // Plot Figura 1.15b - Transitorio forza motrice
    plt::figure(2);
    for (auto& [params, C] : transitorioFmotrice)
        plt::plot(time, C, {{"label", "μ = " + std::to_string(params.first) + ", ν = " + std::to_string(params.second)}});
    plt::xlabel("Tempo (t)");
    plt::ylabel("Forza motrice Fm");
    plt::title("Transitorio della forza motrice - C.I. nulle");
    plt::legend();
    plt::grid(true);

    plt::show();

    return 0;
}

File: capitolo_1/octave/Esempio1_17.m
function Esempio1_17()
  close all
  clear all
  clc
  % Si voglia implementare uno script che secondo l'esempio riporduca i grafici
  % della Figura 1.15a e 1.15b

  %
  % Dati 1.15: Processo massa-molla nominale sotto l'azione del controllore:
  %   Fm(t) = Fmc''(t) = mu * e(t) + nu * int_0^t (e(tau)dtau) , mu>0, nu>0
  % Si rappresentino:
  %   a) transitorio della posizione
  %   b) transitorio della forza motrice
  %
  Mbar = 1; %Volore scelto di progetto
  hbar = 3; %Volore scelto di progetto
  kbar = 1; %Volore scelto di progetto
  sbar = 1; %Volore scelto di progetto
  Fe = 0; %Ipotesi in cui Fe = Febar
  s0 = 0; %Condizioni iniziali nulle sulla posizione
  s0dot = 0; %Condizioni iniziali nulle sulla velocità
  int0 = 0; %Condizioni iniziali nulle sull'integrale d'errore
  x0 = [s0 s0dot int0]; %Vettore condizioni iniziali
  dt = 0.001; %Definizione del tempo
  t = 0:dt:10;
  mus = [1 1.5 3]; % Valori di mu per fare il grafico
  nus = [0.5 1]; % Valori di nu per fare il grafico

  % Figura 1.15a
  transitorioPosizione = zeros([length(t), length(mus)*length(nus)]);
  transitorioFm = zeros([length(t), length(mus)*length(nus)]);
  idx = 1;
  legends = {};
  for i = 1 : length(mus)
    for j = 1 : length(nus)
      [t,x] = ode45(@controlloreAnelloChiuso, t, [x0], Mbar, hbar, kbar, Fe, sbar, mus(i), nus(j));
      transitorioPosizione(:,idx) = x(:,1);
      transitorioFm (:,idx) = mus(i) * sbar - mus(i) * x(:,1) + nus(j) * x(:,3);
      legends{idx} = ['\mu = ' num2str(mus(i)) ', \nu = ' num2str(nus(j))];
      idx = idx + 1;
    endfor
  endfor

  figure(1)
  plot(t,transitorioPosizione, [0 10], [1 1], 'k');
  grid;
  xlabel('Tempo (t)')
  ylabel('Stato s')
  title('Evoluzione dello stato s per diversi guadagni (\mu, \nu) e C.I. nulle')
  legend(legends,'Location','SouthEast');

  figure(2)
  plot(t,transitorioFm, [0 10], [0 0], 'k');
  grid;
  xlabel('Tempo (t)')
  ylabel('Forza motrice')
  title('Evoluzione della Fm per diversi guadagni (\mu, \nu) e C.I. nulle')
  legend(legends,'Location','SouthEast');

endfunction

function [xp] = controlloreAnelloChiuso(t, x, M, h, k, Fe, s, mu, nu)
  % Questa funzione implementa l'evoluzione di stato per l'esempio descritto.
  % Data l'equazione
  %   M * sddot(t) = -h * sdot(t) - k * s(t) + mu * e(t) +
  %                   nu * int_0^t [e(tau)dtau] - Fe
  % Si considera inoltre e = sbar - s e lo stato composto da
  %   x(t) = [ s(t)  sdot(t) int_0^t]
  % Pertanto l'evoluzione di stato sarà
  %   xdot(t) = [ sdot(t)
  %               ((-h/M)*sdot(t) - (k/M)*s(t) + (mu/M)*e(t) + (nu/M)*int_0^t - Fe/M
  %               e(t)]
  e = s - x(1);
  xp = [x(2) ...
        -(h/M) * x(2) - (k/M) * x(1) + (mu/M) * e + (nu/M) * x(3) - Fe/M ...
        e];
endfunction
